In [1]:
#Segmenting and Clustering Neighborhoods in Toronto

## Question 1
### Importing Data

In [2]:
import pandas as pd
import requests

In [4]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

In [5]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [6]:
len(wiki_data),type(wiki_data)

(3, list)

##### We need the first table alone, so dropping the other tables

In [7]:
wiki_data = wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

###### Dropping Borough which are not assigned

In [8]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

##### Grouping the records based on Postal Code

In [9]:
df = df.groupby(['Postal Code']).head()
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

##### Checking for number of records where Neighbourhood is "Not assigned"

In [10]:
df.Neighbourhood.str.count("Not assigned").sum()

0

In [11]:
df = df.reset_index()
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

In [12]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [13]:
df.shape

(103, 3)

#### Answer Question 1: We have 103 rows and 3 Columns

### Question 2
#### Installing Geocoder

In [14]:
! pip install geocoder

     |████████████████████████████████| 98 kB 4.7 MB/s eta 0:00:011


In [15]:
import geocoder # import geocoder

In [17]:
# initialize your variable to None
#lat_lng_coords = None
#ostal_code = "M3A"

# loop until you get the coordinates
#while(lat_lng_coords is None):
 # g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
  #lat_lng_coords = g.latlng

#latitude = lat_lng_coords[0]
#longitude = lat_lng_coords[1]

#### The above package is taking longer time and , not able to get the geographical coordinates . So as suggested in the assignment importing the CSV files from the URL

In [18]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [19]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", data.shape)


The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


##### Since the dimensions are the same, we can try to join on the postal codes to get the required data.

##### Checking the column types of both the dataframes, especially Postal Code column since we are trying to join on it

In [22]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [23]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [26]:
combined_data = df.join(data.set_index('Postal Code'), on = 'Postal Code', how = 'inner')
combined_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [27]:
combined_data.shape

(103, 5)

#### Answer: We get 103 rows as expected when we do a inner join

### Question 3
#####  clustering Toronto based on the similarities of the venues categories using Kmeans clustering and Foursquare API.

In [29]:
!pip install geopy
from geopy.geocoders import Nominatim

     |████████████████████████████████| 112 kB 7.0 MB/s eta 0:00:01


In [30]:
address = 'Toronto, Ontario'
geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


##### Visulization of the map of Toronto

In [31]:
import folium

In [32]:
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

In [33]:
CLIENT_ID = 'EX2BNQTN2XIVF1UZY2CX5T13AHTW5BNGQEVHZABLYY3JZPC5' 
CLIENT_SECRET = 'KC0VSTIG5QAT1YXSW3GQP5PFDFIEPA3NCVXWUOXTEP43OOQ3'
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: EX2BNQTN2XIVF1UZY2CX5T13AHTW5BNGQEVHZABLYY3JZPC5
CLIENT_SECRET:KC0VSTIG5QAT1YXSW3GQP5PFDFIEPA3NCVXWUOXTEP43OOQ3


#####  create a function to get all the venue categories in Toronto

In [34]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

##### Collecting the venues in Toronto for each Neighbourhood

In [35]:
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])


Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [36]:
venues_in_toronto.shape

(1325, 5)

####  1325 records and 5 columns. Checking sample data

In [37]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2  Victoria Village               43.725882               -79.315572   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1           Variety Store      Food & Drink Shop  
2  Victoria Village Arena           Hockey Arena  
3               Portugril  Portuguese Restaurant  
4             Tim Hortons            Coffee Shop

##### Checking the Venues based on Neighbourhood

In [38]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                      Victoria Village   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1312  Mimico NW, The Queensway West, South of Bloor,...   
1313  Mimico NW, The Queensway West, South of Bloor,...   
1314  Mimico NW, The Queensway West, South of Bloor,...   
1315  Mimico NW, The Queensway West, South of Bloor,...   
1316  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                  43.753259               -79.329656         Brookbanks Park   
1                  43.753259               -79.329656           Variety Store   
2                  43.725882               -79.315572  Victoria Village Arena   
3                  43.725882               -79.315572               Portugril   
4                  43.725882               -79.315572             Tim Hortons   
...                      ...                      ...                     ...   
1312               43.628841               -79.520999              Wingporium   
1313               43.628841               -79.520999        South St. Burger   
1314               43.628841               -79.520999               Dollarama   
1315               43.628841               -79.520999          Healthy Planet   
1316               43.628841               -79.520999                7-Eleven   

             Venue Category  
0                      Park  
1         Food & Drink Shop  
2              Hockey Arena  
3     Portuguese Restaurant  
4               Coffee Shop  
...                     ...  
1312            Wings Joint  
1313           Burger Joint  
1314         Discount Store  
1315        Supplement Shop  
1316      Convenience Store  

[410 rows x 5 columns]

#### There are 410 records for each neighbourhood.

#### Checking for the maximum venue categories

In [40]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Accessories Store                                   Wexford, Maryvale   
Adult Boutique                                   Church and Wellesley   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Lounge      CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Warehouse Store                                      Thorncliffe Park   
Wine Bar                                     Little Portugal, Trinity   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                                     Caledonia-Fairbanks   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.750072               -79.295849   
Adult Boutique                   43.665860               -79.383160   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Lounge                   43.628947               -79.394420   
...                                    ...                      ...   
Warehouse Store                  43.705369               -79.349372   
Wine Bar                         43.653206               -79.400049   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.689026               -79.453512   
Yoga Studio                      43.715383               -79.321558   

                                              Venue  
Venue Category                                       
Accessories Store                       Puffin Gear  
Adult Boutique                            Seduction  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Lounge                        Porter Lounge  
...                                             ...  
Warehouse Store                              Costco  
Wine Bar                            Paris Paris Bar  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[234 rows x 4 columns]

In [ ]:
#### There are 234 different types of Venue Categories! 

### One Hot encoding the venue Categories

In [41]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Adult Boutique  Airport  Airport Food Court  \
0                     0               0        0                   0   
1                     0               0        0                   0   
2                     0               0        0                   0   
3                     0               0        0                   0   
4                     0               0        0                   0   
...                 ...             ...      ...                 ...   
1320                  0               0        0                   0   
1321                  0               0        0                   0   
1322                  0               0        0                   0   
1323                  0               0        0                   0   
1324                  0               0        0                   0   

      Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0                  0                0                 0                    0   
1                  0                0                 0                    0   
2                  0                0                 0                    0   
3                  0                0                 0                    0   
4                  0                0                 0                    0   
...              ...              ...               ...                  ...   
1320               0                0                 0                    0   
1321               0                0                 0                    0   
1322               0                0                 0                    0   
1323               0                0                 0                    0   
1324               0                0                 0                    0   

      Antique Shop  Aquarium  ...  Trail  Train Station  \
0                0         0  ...      0              0   
1                0         0  ...      0              0   
2                0         0  ...      0              0   
3                0         0  ...      0              0   
4                0         0  ...      0              0   
...            ...       ...  ...    ...            ...   
1320             0         0  ...      0              0   
1321             0         0  ...      0              0   
1322             0         0  ...      0              0   
1323             0         0  ...      0              0   
1324             0         0  ...      0              0   

      Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                                 0                 0                      0   
1                                 0                 0                      0   
2                                 0                 0                      0   
3                                 0                 0                      0   
4                                 0                 0                      0   
...                             ...               ...                    ...   
1320                              0                 0                      0   
1321                              0                 0                      0   
1322                              0                 0                      0   
1323                              0                 0                      0   
1324                              0                 0                      0   

      Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                   0         0            0              0            0  
1                   0         0            0              0            0  
2                   0         0            0              0            0  
3                   0         0            0              0            0  
4                   0         0            0              0            0  
...               ...       ...          ...            ... 

#####  Adding the neighbourhood to the encoded dataframe

In [42]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  Adult Boutique  Airport  \
0         Parkwoods                  0               0        0   
1         Parkwoods                  0               0        0   
2  Victoria Village                  0               0        0   
3  Victoria Village                  0               0        0   
4  Victoria Village                  0               0        0   

   Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0                   0               0                0                 0   
1                   0               0                0                 0   
2                   0               0                0                 0   
3                   0               0                0                 0   
4                   0               0                0                 0   

   American Restaurant  Antique Shop  ...  Trail  Train Station  \
0                    0             0  ...      0              0   
1                    0             0  ...      0              0   
2                    0             0  ...      0              0   
3                    0             0  ...      0              0   
4                    0             0  ...      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                0         0            0              0            0  
1                0         0            0              0            0  
2                0         0            0              0            0  
3                0         0            0              0            0  
4                0         0            0              0            0  

[5 rows x 235 columns]

#### We will group the Neighbourhoods, calculate the mean venue categories in each Neighbourhood

In [43]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Adult Boutique  Airport  Airport Food Court  Airport Lounge  \
0             0.0      0.0                 0.0             0.0   
1             0.0      0.0                 0.0             0.0   
2             0.0      0.0                 0.0             0.0   
3             0.0      0.0                 0.0             0.0   
4             0.0      0.0                 0.0             0.0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  ...  \
0              0.0               0.0             0.000000           0.0  ...   
1              0.0               0.0             0.000000           0.0  ...   
2              0.0               0.0             0.000000           0.0  ...   
3              0.0               0.0             0.000000           0.0  ...   
4              0.0               0.0             0.043478           0.0  ...   

   Trail  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0    0.0            0.0                            0.0               0.0   
1    0.0            0.0                            0.0               0.0   
2    0.0            0.0                            0.0               0.0   
3    0.0            0.0                            0.0               0.0   
4    0.0            0.0                            0.0               0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                    0.0              0.0       0.0          0.0   
1                    0.0              0.0       0.0          0.0   
2                    0.0              0.0       0.0          0.0   
3                    0.0              0.0       0.0          0.0   
4                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 235 columns]

##### Function to get the top most common venue categories

In [45]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [46]:

import numpy as np

#### There are way too many venue categories, we can take the top 15 to cluster the neighbourhoods

In [47]:
num_top_venues = 15

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood      1st Most Common Venue  \
0                                        Agincourt  Latin American Restaurant   
1                           Alderwood, Long Branch                Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                       Bank   
3                                  Bayview Village                       Café   
4                Bedford Park, Lawrence Manor East                Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Lounge          Skating Rink        Breakfast Spot   
1           Coffee Shop        Sandwich Place          Skating Rink   
2           Coffee Shop           Gas Station        Ice Cream Shop   
3                  Bank    Chinese Restaurant   Japanese Restaurant   
4        Sandwich Place    Italian Restaurant      Greek Restaurant   

  5th Most Common Venue        6th Most Common Venue  \
0     Electronics Store  Eastern European Restaurant   
1                   Gym           Athletics & Sports   
2      Sushi Restaurant                   Restaurant   
3         Deli / Bodega            Electronics Store   
4      Sushi Restaurant      Comfort Food Restaurant   

         7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0                    Drugstore            Donut Shop               Dog Run   
1                          Pub         Deli / Bodega      Department Store   
2                Shopping Mall          Intersection         Deli / Bodega   
3  Eastern European Restaurant             Drugstore            Donut Shop   
4                     Pharmacy           Pizza Place                   Pub   

      10th Most Common Venue 11th Most Common Venue 12th Most Common Venue  \
0        Distribution Center         Discount Store                  Diner   
1               Dessert Shop                  Diner     Dim Sum Restaurant   
2  Middle Eastern Restaurant      Mobile Phone Shop            Supermarket   
3                    Dog Run    Distribution Center         Discount Store   
4                       Café                Butcher             Restaurant   

  13th Most Common Venue 14th Most Common Venue 15th Most Common Venue  
0     Dim Sum Restaurant           Dessert Shop       Department Store  
1            Curling Ice         Discount Store    Distribution Center  
2            Pizza Place               Pharmacy    Fried Chicken Joint  
3                  Diner     Dim Sum Restaurant           Dessert Shop  
4           Liquor Store         Breakfast Spot      Indian Restaurant

####  Let's make the model to cluster our Neighbourhoods

In [48]:
from sklearn.cluster import KMeans

In [49]:
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

##### Checking the labelling of our model

In [50]:
kmeans.labels_[0:100]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 0, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 4, 1,
       3, 1, 1, 1, 1, 1, 0, 1], dtype=int32)

##### Let's add the clustering Label column to the top 15 common venue categories

In [52]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

##### Join Toronto_grouped with combined_data on neighbourhood to add latitude & longitude for each neighborhood to prepare it for plotting

In [53]:
Toronto_merged = combined_data

Toronto_merged = Toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

Toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels  1st Most Common Venue  \
0  43.753259 -79.329656             0.0                   Park   
1  43.725882 -79.315572             1.0  Portuguese Restaurant   
2  43.654260 -79.360636             1.0            Coffee Shop   
3  43.718518 -79.464763             1.0         Clothing Store   
4  43.662301 -79.389494             1.0            Coffee Shop   

   2nd Most Common Venue   3rd Most Common Venue 4th Most Common Venue  ...  \
0      Food & Drink Shop             Yoga Studio          Dance Studio  ...   
1            Pizza Place            Intersection           Coffee Shop  ...   
2                   Park                    Café                Bakery  ...   
3  Vietnamese Restaurant  Furniture / Home Store    Miscellaneous Shop  ...   
4       Sushi Restaurant             Yoga Studio          Burger Joint  ...   

        6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                   Drugstore            Donut Shop               Dog Run   
1  Financial or Legal Service                 Diner      Department Store   
2                     Theater            Restaurant                   Pub   
3           Accessories Store           Event Space             Gift Shop   
4          Italian Restaurant              Beer Bar   Japanese Restaurant   

  9th Most Common Venue 10th Most Common Venue 11th Most Common Venue  \
0   Distribution Center         Discount Store                  Diner   
1          Dessert Shop     Dim Sum Restaurant            Yoga Studio   
2        Farmers Market            Event Space         Chocolate Shop   
3              Boutique             Donut Shop                Dog Run   
4        Sandwich Place    Distribution Center                  Diner   

  12th Most Common Venue 13th Most Common Venue 14th Most Common Venue  \
0     Dim Sum Restaurant           Dessert Shop       Department Store   
1           Dance Studio         Discount Store    Distribution Center   
2      French Restaurant            Yoga Studio           Dessert Shop   
3    Distribution Center         Discount Store                  Diner   
4          Burrito Place                   Café     Mexican Restaurant   

  15th Most Common Venue  
0          Deli / Bodega  
1                Dog Run  
2                    Spa  
3     Dim Sum Restaurant  
4               Creperie  

[5 rows x 21 columns]

##### Drop all the NaN values to prevent data skew

In [54]:
Toronto_merged_nonan =Toronto_merged.dropna(subset=['Cluster Labels'])

##### Plotting the clusters on the map

In [56]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [57]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged_nonan['Latitude'], Toronto_merged_nonan['Longitude'], Toronto_merged_nonan['Neighbourhood'], Toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

#### Let's verify each of our clusters

#### Cluster 1

In [60]:

Toronto_merged_nonan.loc[Toronto_merged_nonan['Cluster Labels'] == 0, Toronto_merged_nonan.columns[[1] + list(range(5, Toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York             0.0                  Park   
21              York             0.0                  Park   
35         East York             0.0                  Park   
61   Central Toronto             0.0                  Park   
66        North York             0.0                  Park   
85       Scarborough             0.0                  Park   
91  Downtown Toronto             0.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0      Food & Drink Shop           Yoga Studio          Dance Studio   
21         Women's Store                  Pool          Dance Studio   
35     Convenience Store           Yoga Studio         Deli / Bodega   
61           Swim School              Bus Line           Yoga Studio   
66     Convenience Store           Yoga Studio         Deli / Bodega   
85            Playground           Coffee Shop          Intersection   
91            Playground                 Trail           Yoga Studio   

          5th Most Common Venue        6th Most Common Venue  \
0   Eastern European Restaurant                    Drugstore   
21  Eastern European Restaurant                    Drugstore   
35            Electronics Store  Eastern European Restaurant   
61                Deli / Bodega  Eastern European Restaurant   
66            Electronics Store  Eastern European Restaurant   
85                  Yoga Studio                 Dance Studio   
91                  Curling Ice                    Drugstore   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0             Donut Shop               Dog Run   Distribution Center   
21            Donut Shop               Dog Run   Distribution Center   
35             Drugstore            Donut Shop               Dog Run   
61             Drugstore            Donut Shop               Dog Run   
66             Drugstore            Donut Shop               Dog Run   
85             Drugstore            Donut Shop               Dog Run   
91            Donut Shop               Dog Run   Distribution Center   

   10th Most Common Venue 11th Most Common Venue 12th Most Common Venue  \
0          Discount Store                  Diner     Dim Sum Restaurant   
21         Discount Store                  Diner     Dim Sum Restaurant   
35    Distribution Center         Discount Store                  Diner   
61    Distribution Center         Discount Store                  Diner   
66    Distribution Center         Discount Store                  Diner   
85    Distribution Center         Discount Store                  Diner   
91         Discount Store                  Diner     Dim Sum Restaurant   

   13th Most Common Venue 14th Most Common Venue 15th Most Common Venue  
0            Dessert Shop       Department Store          Deli / Bodega  
21           Dessert Shop       Department Store          Deli / Bodega  
35     Dim Sum Restaurant           Dessert Shop       Department Store  
61     Dim Sum Restaurant           Dessert Shop       Department Store  
66     Dim Sum Restaurant           Dessert Shop       Department Store  
85     Dim Sum Restaurant           Dessert Shop       Department Store  
91           Dessert Shop       Department Store          Deli / Bodega

##### Cluster 2

In [59]:
Toronto_merged_nonan.loc[Toronto_merged_nonan['Cluster Labels'] == 1, Toronto_merged_nonan.columns[[1] + list(range(5, Toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels  1st Most Common Venue  \
1          North York             1.0  Portuguese Restaurant   
2    Downtown Toronto             1.0            Coffee Shop   
3          North York             1.0         Clothing Store   
4    Downtown Toronto             1.0            Coffee Shop   
6         Scarborough             1.0   Fast Food Restaurant   
..                ...             ...                    ...   
97   Downtown Toronto             1.0                   Café   
98          Etobicoke             1.0                  River   
99   Downtown Toronto             1.0       Sushi Restaurant   
100      East Toronto             1.0     Light Rail Station   
102         Etobicoke             1.0         Discount Store   

     2nd Most Common Venue   3rd Most Common Venue  \
1              Pizza Place            Intersection   
2                     Park                    Café   
3    Vietnamese Restaurant  Furniture / Home Store   
4         Sushi Restaurant             Yoga Studio   
6              Yoga Studio           Deli / Bodega   
..                     ...                     ...   
97             Coffee Shop              Restaurant   
98                    Pool             Curling Ice   
99             Coffee Shop                Beer Bar   
100            Yoga Studio           Auto Workshop   
102            Wings Joint                  Bakery   

           4th Most Common Venue        5th Most Common Venue  \
1                    Coffee Shop                 Hockey Arena   
2                         Bakery               Breakfast Spot   
3             Miscellaneous Shop                  Coffee Shop   
4                   Burger Joint                          Bar   
6              Electronics Store  Eastern European Restaurant   
..                           ...                          ...   
97            Seafood Restaurant                    Gastropub   
98   Eastern European Restaurant                    Drugstore   
99                          Café                          Pub   
100                         Park                   Smoke Shop   
102               Sandwich Place                Grocery Store   

          6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
1    Financial or Legal Service                 Diner      Department Store   
2                       Theater            Restaurant                   Pub   
3             Accessories Store           Event Space             Gift Shop   
4            Italian Restaurant              Beer Bar   Japanese Restaurant   
6                     Drugstore            Donut Shop               Dog Run   
..                          ...                   ...                   ...   
97             Asian Restaurant                 Hotel            Steakhouse   
98                   Donut Shop               Dog Run   Distribution Center   
99             Ramen Restaurant          Dessert Shop  Ethiopian Restaurant   
100               Garden Center  Gym / Fitness Center                Garden   
102                         Gym         Tanning Salon  Fast Food Restaurant   

    9th Most Common Venue 10th Most Common Venue  11th Most Common Venue  \
1            Dessert Shop     Dim Sum Restaurant             Yoga Studio   
2          Farmers Market            Event Space          Chocolate Shop   
3                Boutique             Donut Shop                 Dog Run   
4          Sandwich Place    Distribution Center                   Diner   
6     Distribution Center         Discount Store                   Diner   
..                    ...                    ...                     ...   
97                    Pub                 Bakery          Sandwich Place   
98         Discount Store                  Diner      Dim Sum Restaurant   
99      Indian Restaurant            Escape Room                    Park   
100               Brewery            Pizza Place              Comic Shop   
102       Supplement Shop     

##### Cluster 3

In [61]:
Toronto_merged_nonan.loc[Toronto_merged_nonan['Cluster Labels'] == 2, Toronto_merged_nonan.columns[[1] + list(range(5, Toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
57   North York             2.0        Baseball Field           Yoga Studio   
101   Etobicoke             2.0        Baseball Field           Yoga Studio   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
57            Event Space           Escape Room     Electronics Store   
101           Event Space           Escape Room     Electronics Store   

           6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
57   Eastern European Restaurant             Drugstore            Donut Shop   
101  Eastern European Restaurant             Drugstore            Donut Shop   

    9th Most Common Venue 10th Most Common Venue 11th Most Common Venue  \
57                Dog Run    Distribution Center         Discount Store   
101               Dog Run    Distribution Center         Discount Store   

    12th Most Common Venue 13th Most Common Venue 14th Most Common Venue  \
57                   Diner     Dim Sum Restaurant           Dessert Shop   
101                  Diner     Dim Sum Restaurant           Dessert Shop   

    15th Most Common Venue  
57        Department Store  
101       Department Store

##### Cluster 4

In [62]:
Toronto_merged_nonan.loc[Toronto_merged_nonan['Cluster Labels'] == 3, Toronto_merged_nonan.columns[[1] + list(range(5, Toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
64    York             3.0     Convenience Store           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
64         Deli / Bodega           Escape Room     Electronics Store   

          6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
64  Eastern European Restaurant             Drugstore            Donut Shop   

   9th Most Common Venue 10th Most Common Venue 11th Most Common Venue  \
64               Dog Run    Distribution Center         Discount Store   

   12th Most Common Venue 13th Most Common Venue 14th Most Common Venue  \
64                  Diner     Dim Sum Restaurant           Dessert Shop   

   15th Most Common Venue  
64       Department Store

##### Cluster 5

In [63]:
Toronto_merged_nonan.loc[Toronto_merged_nonan['Cluster Labels'] == 4, Toronto_merged_nonan.columns[[1] + list(range(5, Toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
11  Etobicoke             4.0                Bakery           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
11           Event Space           Escape Room     Electronics Store   

          6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11  Eastern European Restaurant             Drugstore            Donut Shop   

   9th Most Common Venue 10th Most Common Venue 11th Most Common Venue  \
11               Dog Run    Distribution Center         Discount Store   

   12th Most Common Venue 13th Most Common Venue 14th Most Common Venue  \
11                  Diner     Dim Sum Restaurant           Dessert Shop   

   15th Most Common Venue  
11       Department Store

#### Clustering of  Toronto neighbourhood based on venue categories was done sucessfully!